From 50b256506b871f5babc504a08855b72396ed60b8 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Tue, 24 Jun 2025 12:06:51 +0800
Subject: [PATCH 101/115] mtk: mac80211: fix kernel crash during reboot &
 cfg80211_autodisconnect_wk

When cfg80211_autodisconnect_wk is executed during the reboot process, a race
condition might occur, causing the link->conf to be NULL in
__ieee80211_sta_cap_rx_bw

 Unable to handle kernel read from unreadable memory at virtual address 0000000000000088
 Mem abort info:
   ESR = 0x0000000096000005
   EC = 0x25: DABT (current EL), IL = 32 bits
   SET = 0, FnV = 0
   EA = 0, S1PTW = 0
   FSC = 0x05: level 1 translation fault
 Data abort info:
   ISV = 0, ISS = 0x00000005, ISS2 = 0x00000000
   CM = 0, WnR = 0, TnD = 0, TagAccess = 0
   GCS = 0, Overlay = 0, DirtyBit = 0, Xs = 0
 user pgtable: 4k pages, 39-bit VAs, pgdp=0000000042059000
 [0000000000000088] pgd=0000000000000000, p4d=0000000000000000, pud=0000000000000000
  Internal error: Oops: 0000000096000005 [#1] SMP
 Modules linked in: pppoe ppp_async nft_fib_inet l2tp_ppp pptp pppox ppp_mppe ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_flow_table_netlink nft_flow_offload nf_flow_table_inet nf_tables mt7996e(O) mt76_connac_lib(O) mt76(O) mac80211(O) iptable_nat ipt_REJECT cfg80211(O) xt_time xt_tcpudp xt_tcpmss xt_statistic xt_state xt_recent xt_quota xt_pkttype xt_owner xt_nat xt_multiport xt_mark xt_mac xt_limit xt_length xt_hl xt_helper xt_ecn xt_dscp xt_conntrack xt_connmark xt_connlimit xt_connbytes xt_comment xt_cgroup xt_addrtype xt_TCPMSS xt_REDIRECT xt_MASQUERADE xt_LOG xt_HL xt_FLOWOFFLOAD xt_DSCP xt_CLASSIFY ums_usbat ums_sddr55 ums_sddr09 ums_karma ums_jumpshot ums_isd200 ums_freecom ums_datafab ums_cypress ums_alauda spidev slhc sfp rtc_pcf8563 nfnetlink nf_reject_ipv4 nf_nat nf_log_syslog nf_conncount mdio_i2c libcrc32c iptable_mangle iptable_filter
  ipt_ECN ip_tables compat(O) br_netfilter at24 mt76qos(O) crypto_safexcel ntfs3 sg pwm_fan i2c_gpio i2c_algo_bit i2c_mux_pca954x i2c_mux ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 nls_ucs2_utils cifs_arc4 asn1_decoder nfsd msdos ip_gre gre l2tp_netlink l2tp_core udp_tunnel ip6_udp_tunnel ip_tunnel rpcsec_gss_krb5 auth_rpcgss oid_registry lockd sunrpc grace autofs4 nls_utf8 nls_iso8859_1 nls_cp437 marvell10g marvell crypto_user algif_skcipher algif_rng algif_hash algif_aead af_alg sha512_arm64 sha1_ce sha1_generic seqiv md5 geniv des_generic libdes cts cbc authencesn authenc arc4 uas usb_storage leds_gpio xhci_plat_hcd xhci_pci xhci_mtk_hcd xhci_hcd gpio_button_hotplug(O) vfat fat exfat usbcore usb_common aquantia an8801 [last unloaded: ksmbd]
 CPU: 2 PID: 250 Comm: kworker/2:2 Tainted: G           O       6.6.92 #0
 Hardware name: MediaTek MT7987A RFB (DT)
 Workqueue: events cfg80211_autodisconnect_wk [cfg80211]
 pstate: 60400005 (nZCv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
 pc : __ieee80211_sta_cap_rx_bw+0xb4/0xcc [mac80211]
 lr : _ieee80211_sta_cap_rx_bw+0x10/0x24 [mac80211]
 sp : ffffffc0817838e0
 x29: ffffffc0817838e0 x28: 0000000000000000 x27: 0000000000000000
 x26: ffffff8006c7e678 x25: 0000000000000000 x24: ffffff8010951ad0
 x23: ffffff8010950900 x22: ffffff800a378600 x21: ffffff8010950900
 x20: ffffff8006c7e000 x19: ffffff8010951a58 x18: 0000000000000468
 x17: 662030206b6e696c x16: 2065766f6d657220 x15: ffffffc081783af0
 x14: ffffff8010835c18 x13: 0000000000000468 x12: 00000000ffffffea
 x11: 00000000ffffefff x10: ffffffc080e1d890 x9 : ffffffc080dc5838
 x8 : ffffffc081783b70 x7 : 0000000000000000 x6 : 0000000000000000
 x5 : 0000000000000fff x4 : ffffff8006c7e678 x3 : 00000000000015f0
 x2 : ffffff8006c7eb60 x1 : ffffff8010844900 x0 : 0000000000000000
 Call trace:
  __ieee80211_sta_cap_rx_bw+0xb4/0xcc [mac80211]
  _ieee80211_recalc_chanctx_min_def+0x1a8/0x3ec [mac80211]
  ieee80211_recalc_chanctx_min_def+0x1c/0x6c [mac80211]
  _ieee80211_change_chanctx+0x2cc/0x394 [mac80211]
  ieee80211_recalc_chanctx_chantype+0x180/0x1dc [mac80211]
  ieee80211_assign_link_chanctx+0x27c/0x4b4 [mac80211]
  __ieee80211_link_release_channel+0x60/0x164 [mac80211]
  ieee80211_link_release_channel+0x1c/0x28 [mac80211]
  ieee80211_stop_ap+0x1d4/0x378 [mac80211]
  ___cfg80211_stop_ap+0xf4/0x334 [cfg80211]
  cfg80211_stop_ap+0x54/0xd4 [cfg80211]
  cfg80211_autodisconnect_wk+0xc0/0x128 [cfg80211]
  process_one_work+0x17c/0x3bc
  worker_thread+0x2e8/0x4d0
  kthread+0xd8/0xdc
  ret_from_fork+0x10/0x20
 Code: d282be03 8b030000 8b000020 f9400800 (f9404400)
 ---[ end trace 0000000000000000 ]---
 pstore: backend (ramoops) writing error (-28)
 Kernel panic - not syncing: Oops: Fatal exception
 SMP: stopping secondary CPUs
 Kernel Offset: disabled
 CPU features: 0x0,00000000,00000000,1000400b
 Memory Limit: none
 Starting Memory dump SMCC
 Memory dump SMCC failed
 Rebooting in 3 seconds..

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/vht.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/vht.c b/net/mac80211/vht.c
index a3464c49..befc6f10 100644
--- a/net/mac80211/vht.c
+++ b/net/mac80211/vht.c
@@ -372,6 +372,10 @@ __ieee80211_sta_cap_rx_bw(struct link_sta_info *link_sta,
 
 			rcu_read_lock();
 			link_conf = rcu_dereference(sdata->vif.link_conf[link_id]);
+			if (unlikely(!link_conf)) {
+				rcu_read_unlock();
+				return IEEE80211_STA_RX_BW_20;
+			}
 			band = link_conf->chanreq.oper.chan->band;
 			rcu_read_unlock();
 		}
-- 
2.45.2

